orttraining/orttraining/core/session/training_session.cc

```diff
@@ -155,8 +155,15 @@ Status TrainingSession::ConfigureForTraining(
   is_mixed_precision_enabled_ = config.mixed_precision_config.has_value();
 
   std::string loss_name{};
+  // Enable loss scale if mixed precision is enabled AND at pipeline last stage if pipeline is used.
+  // We are currently making the assumption that no data parallelism is used together with model parallelism.
+  // So we can check the last stage by checking the world_rank and world_size. Once DP and MP combination is
+  // enabled, we need to device another way to check MP stages.
+  bool enable_loss_scale = is_mixed_precision_enabled_ &&
+                           (!config.pipeline_config.has_value() ||
+                            (config.distributed_config.world_rank + 1 == config.distributed_config.world_size));
```

Comment:
>config.distributed_config.world_rank + 1 == config.distributed_config.world_size [](start = 29, length = 80)

Seems this have some hardcoded assumption on D/H/P partitioning. Do we have a util function for distributed_config to tell if this is last stage of pipeline?

Reply:
Unfortunately no. Without introducing a new worker group, there won't be a systematic formular to deduce a certain rank's position in D/H/P. Currently we run pipeline without DP, and this assumption is scattered in a few places. I am not too worried about it though, as it always comes after a check of whether pipeline is used. In the future when we have a new util function, search and replace should cover those places.

---
In reply to: [434895205](https://github.com/microsoft/onnxruntime/pull/4113#discussion_r434895205) [](ancestors = 434895205)